Begin AAA Dashboard Backend

emord opened the pull request:
This is a big PR but that's mostly because of the UCR data source definitions. I'm working on getting a test dataset up for automated tests, but just finished the first aggregation of data.

This is a fairly large departure from the CAS dashboard schema. I've decided this is the correct path forward to solve some real issues we've encountered. Some differences include:

* No similar tables to child_health_monthly or ccs_record_monthly. Instead we will have a Woman, ccs_record, and Child models. These models will store all the current information of the case and some key information used for date filters (such as ranges of time that a woman was pregnant) (not in this PR, but there will be similar tables to the rollup tables like agg_awc.
* Detailed historical information will be stored in array fields that include the changes a case property has seen over time. These should only be used for detailed beneficiary information or background tasks. This was inspired by a limitation of the CAS dashboard where we can only display the past 6 months of a child's measurements because of performance issues with querying a different table for every month
* No large aggregation step. Each aggregation will only update with information that has come in since the previous aggregation. Its very similar to https://www.citusdata.com/blog/2018/06/14/scalable-incremental-data-aggregation/ and I would suggest reading that blog post to understand the workflow that is being attempted.
* I have abstracted to a Woman model as a Woman in this program can flow between pregnant and "eligible couple" It felt easiest to have one data source for a person.

review commit by commit. There is a lot more iteration that will be going on with this work. And I'd like to keep the PRs small. This should have no affect to any other projects and as such should be ok to merge. 

Some more of the work includes:

* populate the woman, ccs_record and history tables
* create aggregate tables for the location levels
* populate a test data set to add to the automated tests
* change the static numbers that were introduced in https://github.com/dimagi/commcare-hq/pull/23110

Likely interested parties: @calellowitz @czue @snopoke 
buddies @nickpell @millerdev

emord committed:
Add ccs_record case data source for AAA pilot

emord committed:
Add child_health case data source for AAA pilot

emord committed:
Add household case data source for AAA pilot

emord committed:
Add person case data source for AAA pilot

emord committed:
Add tasks case data source for AAA pilot

emord committed:
Add growth monitoring forms for AAA pilot

emord committed:
Start AAA pilot models

emord committed:
Add eligible_couple data source for AAA pilot

emord committed:
Start aggregation work

emord committed:
Add location UCRs for AAA pilot

emord committed:
Add aaa-data database

emord commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+        DateRangeField(),
+        help_text="The ranges in which a ccs_record has been opened and the baby has not been born",
+        null=True
+    )
```
This is not a pattern that we've followed before. I have verified that something similar to the following works:

`'2018-02-01'::date <@ any(array[daterange('2015-01-06','2015-02-10'), daterange('2017-10-05','2017-10-11')])`

There is not currently an accessible way to implement that in django, but I believe it should be feasible by writing a custom Django query expression

snopoke replied:
Similar to above, if this is going to be used for filtering the table, have we done any profiling to see how it performs on large tables?

emord replied:
I have not yet done profiling, but did check that this can be used with a GiST index. Unfortunately those aren't natively supported in django until 2.0, so I'll need to do something a bit more custom in the meantime

emord committed:
Mark migration for no linting

emord committed:
Fix syntax errors in AAA data sources

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+Conventions followed in this file:
+    * Fields ending in _ranges represent a field's history from forms that are useful for filtering.
+      This is only used for fields that are required to filter for specific ranges (usually a date range).
+    * Fields ending in _history represent a field's entire history from forms.
```
Do we know how big these could get? Could be useful to profile.

emord replied:
I am not sure and it will be different for every indicator where the history is stored. For example things like weight and height of a child are measured every month, but an eligible couple's history of family planning methods are likely much smaller.

Another thing that makes this harder is that for large values, these are stored in TOAST tables and not the main table (similar to how large strings are stored). 

These history methods are only intended to be used in background aggregation tasks or in single beneficiary view tables, so there should not be any large user facing queries that need to use them.

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+from corehq.apps.userreports.util import get_table_name
+
+
+class LocationDenormalizedModel(models.Model):
```
Not that I think it applies heavily to these models but just mentioning for possibly future work that optimising column order should be considered for large tables: https://blog.2ndquadrant.com/on-rocks-and-sand/ 

It might actually make more sense for the UCR tables.

emord replied:
Agreed that there's probably more low hanging fruit than this, but up to 20% space savings does seem worth exploring eventually.

I had heard of struct padding in C but never quite made the connection to SQL.

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+    village_id = models.TextField(null=True)
+
+    class Meta(object):
+        abstract = True
```
:+1:

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+        she will have two ccs_records for this pregnancy
+    """
+
+    person_case_id = models.TextField()
```
should this be a foreign key?

emord replied:
I've avoided foreign keys for now as I'm not clear on all the implications with Citus and our workflows there. I wanted to make this with a migration in mind and wanted to keep it simple for that.

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+    # but going with child_health as there's a potential for person case to live
+    # on after the child_health case stops being tracked (into adolescent girl, or woman)
+    opened_on = models.DateField(help_text="child_health.opened_on")
+    closed_on = models.DateField(help_text="child_health.opened_on", null=True)
```
help text should be `closed_on`

snopoke commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,272 @@
+            FROM "{person_cases_ucr_tablename}"
+            WHERE inserted_at >= %(window_start)s AND inserted_at < %(window_end)s
+        ) person
+        WHERE child.person_case_id = person.doc_id
```
Is there a case where the row may not exist in the table yet?

emord replied:
There will be many  person cases that will not exist in the child table.  there does exist a case where:

1) person case is updated
2) aggregation window happens
3) child case updated
4) aggregation window happens
5) person case is not in this inserted_at window.

I think this could be solved in the background with a task that monitors the child table to find records that have these key  columns NULL, or perhaps a two step aggregation where we update person cases that have changed and also update records with NULL columns. I may change this for the pilot to do full data aggregations in the short term while this is more solved

snopoke commented on custom/aaa/tasks.py:
```diff
@@ -0,0 +1,59 @@
+    agg_query, agg_params = method(domain, window_start, window_end)
+    with connections['aaa-data'].cursor() as cursor:
+        cursor.execute(agg_query, agg_params)
+    agg_info.end_time = datetime.utcnow()
```
This seems a little dangerous to me since it could result in missing a few rows that get inserted between the first date and now. Is it possible to have the query return the max `inserted_at` which could be used for the end time?

snopoke reviewed (commented):
Are the UCRs going to be used for anything outside of aggregations? Just wondering if it could be possible to purge old data from the UCRs.

snopoke commented:
+1 on the change in aggregation method!

millerdev commented on custom/aaa/tasks.py:
```diff
@@ -0,0 +1,59 @@
+    else:
+        window_start = window_start[0]
+
+    window_end = datetime.utcnow()
```
Looks like this is using date-based aggregation windows rather than sequence-id based with the intricate locking procedure described in [the citus article](https://www.citusdata.com/blog/2018/06/14/scalable-incremental-data-aggregation/). Did you decide against the approach outlined in that article, or are you using a variant that does not have all the same characteristics?

emord replied:
Yeah. I'm using dates because we insert the current time in inserted_at and we do not have a sequential ID. inserted_at is the only increasing ID we have for now. (the primary key is doc_id which is random uuid)

millerdev replied:
From the article:

> There might still be in-progress writes to the events table that were assigned lower sequence values, but are not yet visible when the aggregation runs

Since this is not using the locking trick, I think the same applies to dates the way they're used here, which means some rows may be missed by the aggregation. Also I think small date differences between servers could cause issues.

emord replied:
Agreed about the small date differences between different dbs making this messy. Its possible that we could define some amount of drift that is expected and have the aggregation window end there. So the end date is always `now() - 1 minute`. That would solve this for machines with drift of up to one minute. Currently this is not an issue because it will all be on one server. It will be good to look into what sort of guarantees Citus can provide about synchronization of nodes (thought currently that utcnow does come from the celery task in python. Possible that we could use `now()` ad defined by postgres).

I am planning to use the locking trick as commented in line 28. From my understanding the workflow would be

1. Set window's end time
2. Lock the table for writes
3. Release lock.
    * all writes to the database that were inserted_at < time at point 1 are guaranteed to have been written.
4. Agg data that has been inserted up until point 1.

millerdev replied:
> I am planning to use the locking trick as commented in line 28.

Ah, I didn't realize that comment was a TODO item.

Using dates rather than a sequence still seems to me more complex and harder to reason about. `datetime.utcnow()` is not monotonically increasing (clock drift is real, and the NTP service adjusts system time routinely). See [SO answer](https://stackoverflow.com/a/3527632/10840) for some examples of how complex a subject this is. If I were doing this I would use a sequence as described in the article.

emord replied:
Yeah. That does make a lot of sense. We don't currently have a way to add sequences to a UCR data source, but I imagine it would not be difficult. I would need to think more about adding it to case UCRs since those are updated and not just inserted (like the 99% case for forms).

I have [removed the where clauses](https://github.com/dimagi/commcare-hq/pull/23243/commits/f702bd968604037b5b24655b89d85f17400f4a1b) for now while I think about this. This means that the aggregation_window_start and window_end are not used and each aggregation is a full aggregation. I have kept those database as they will still be useful concepts for the monthly location rollup tables that are coming next.

emord commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,149 @@
+
+    # Shared location hierarchy
+    state_id = models.TextField(null=True)
+    district_id = models.TextField(null=True)
```
I'm intentionally using denormalization for now. I'm attempting to balance getting something QA-able before the pilot (march 1) vs supporting future CitusDB backend vs supporting future reports vs similarities to the current CAS dashboard so hopefully we can consolidate the custom warehouses and its easier to spin up on reports.

I'm not sure if I have made the right trade-off but I think it was easiest to do this for now and possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs. It's possible that normalizing now with a foreign key to a location table would be more performant and easy to migrate, but with an impending Citus migration I'm less confident that I understand all the moving pieces there.

czue replied:
> possibly in the future have the location hierarchy be a "reference" table that's on every node for efficient JOINs

no data to back this up, but this sounds intuitively like a good option to me, and would be how a standard "star schema" dimensional data warehouse design would tackle the problem.

this is quite a lot of redundant storage assuming one of the low-level ids can deterministically generate the rest.

also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

emord replied:
> also probably obvious, but if a district's state changes, then you have to update a huge amount of row-level data to handle that whereas if location is a dimension then you just update it in relevant rows of the location table which is much smaller.

FWIW This doesn't happen right now, but it could in the future. Before this does happen on a large scale for CAS there should be more speccing and understanding of what they actually want. There's a case to be made that we would not want the districts state to change for historical data as it would point to the geography that was relevant at that time. 

There would be similar questions for beneficiaries moving between AWC or ASHAs. Should we report only the current ASHA, or have the beneficiary report the different ones? 

Not that any of that is actionable right now, but I've heard the talks of this workflow being added for the past year or so, but no progress has been made on actually doing it.

emord commented:
> Are the UCRs going to be used for anything outside of aggregations?

Not at the moment. I haven't been asked for any mobile reports at this time. 

>Just wondering if it could be possible to purge old data from the UCRs.

Most of these are case data UCRs so we can't really purge them (or at least a large percentage of them) as they'll be updated fairly frequently.

For form UCRs time based partitioning makes a lot of sense to make sure that old data does not affect query performance. Purging the data is possible as we usually don't care about any data that's very old, but its possible that we would need to rebuild these if a new report was requested and we lost some of the raw information in the aggregation.

For example if they wanted a report about # of forms of each of the form types submitted over the life of the project (so like # BP forms, # EC forms, # hh forms, etc), but we only saved the total number of forms per month (AWC submitted 7 forms, etc) then we'd need to rebuild all forms. (I realize that for that example we'd have easier ways to do this, but it was the best one I could think of at the time)

emord committed:
Merge branch 'master' into je/aaa-report-backend

emord committed:
Fix closed_on help_text

emord committed:
Aggregte all the data.

Some good points were brought up
https://github.com/dimagi/commcare-hq/pull/23243#discussion_r256728746
and https://github.com/dimagi/commcare-hq/pull/23243#discussion_r256961315
about potential issues aggregating solely based on time.

Due to this feedback the timing aggregation will be brought back after
more thought to better accommodate servers with different times and out
of order case and form processing

This is not a short term concern as we only have 100 users for the pilot
and so each UCR will be less than 100k rows.

emord committed:
Aggregate Woman data

emord committed:
Aggregate CcsRecord data

emord committed:
DRY for village and AWC location insertion

emord committed:
Aggregate pregnancy ranges from ccs records

emord committed:
Stickler

emord commented:
@snopoke @millerdev  Thanks a ton for the feedback today.

I've added some more aggregations to populate the models I have now. There's a bit of repeated code that should be cleaned up (mostly in tasks.py and the SQL could be better generalized with helpers).

Some things that I will be looking into as the pilot goes on:

- Using sequence IDs instead of time (and adding sequence ids to UCR)
- General profiling both with query times and sizes
- Foreign keys & normalization and how that may impact changing from postgres to Citus.
- Curating a test dataset that is able to sufficiently test for the edge cases in the app.

For now I think that it makes sense to keep this base schema, and I will profile it with the feedback you guys have given me in mind which will inform next steps. I'll also keep up to date with Citus work and make changes based on that as needed. Let me know what you think of this approach

snopoke commented:
+1 on the discussion. I think it will be very valuable to come up with an aggregation workflow that we can trust and that we can run often. A few more thoughts from me:

* we could potentially query the DB to get the window end knowing that any data which get's added after that point will have a lower date (not sure if that's 100% true if there are uncommitted transactions that could have a lower date).
  * Doing that on a sharded DB would be more challenging since you'd want to take the min of the dates from all shards

* Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases. In the example they are giving they are sharding on site_id and so never aggregating across sites which means that all aggregation queries run locally on shards. We would need to be able to guarantee that (at least the first level aggregation) can always happen on the local shard database. Something that isn't clear to me in their example is that the rollup table isn't distributed (maybe they just left that out of the example).

millerdev commented:
+1 sounds good to me.

Responding to @snopoke's idea:
> could potentially query the DB to get the window end knowing that any data which get's added after that point will have a lower date

Assuming you meant "higher date" (more recent)—or did I misunderstand what you were saying? I don't think this will work because of things like clock drift and NTP time adjustments.

We might be able to make it work if we only ever aggregate up to some point sufficiently in the past that it will never be overlapped by drift or clock adjustments. Basically selecting a time range into which we can guarantee new data will never be written. However, I still think this is going to be more complicated than using a sequence.

emord commented:
> * Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases

Is possible that we could have this aggregation loop over supervisor ids? Would need to think about how that affects the MoH hierarchy but it doesn't seem unreasonable at the moment. Also fits in with the concept of many, smaller aggregations

millerdev commented:
> * Using sequence IDs definitely seems better but it only works if you can run the aggregations on the shard databases

I was under the impression that a sharded table could share a single sequence between all shards on citus. Is this not true?

emord commented:
Offlined with Simon and Daniel. We're on the same page about using this as a base for future work and profiling to define our next steps and its clear that it wouldn't be immediately scalable tomorrow

czue commented on corehq/sql_db/connections.py:
```diff
@@ -22,6 +22,7 @@
 ICDS_UCR_ENGINE_ID = 'icds-ucr'
 ICDS_UCR_NON_DASHBOARD_ENGINE_ID = 'icds-ucr-non-dashboard'
 ICDS_TEST_UCR_ENGINE_ID = 'icds-test-ucr'
+AAA_DB_ENGINE_ID = 'aaa-data'
```
sorry this is a really dumb question but what is "AAA"?

emord replied:
ASHA, AWW, ANM. Its the more official name of REACH and I didn't want to get into a situation where in 3 years someone asks me why I named this REACH (but I'm sure this won't prevent this)

czue commented on custom/aaa/models.py:
```diff
@@ -0,0 +1,420 @@
+      They are not guaranteed to be ordered by date.
+      They are not intended to be queried on, their interpretation should happen in python.
+      These history fields are saved in a separate model to ensure that the main model stays
+      fast and small for filtering.
```
What's an example of how these `history` fields would be used?

emord replied:
From description:

> * Detailed historical information will be stored in array fields that include the changes a case property has seen over time. These should only be used for detailed beneficiary information or background tasks. This was inspired by a limitation of the CAS dashboard where we can only display the past 6 months of a child's measurements because of performance issues with querying a different table for every month

So instead of generating a query for 60 tables over a child's first 5 years, this should just query the array to return 60 values detailing how it has grown

czue reviewed (commented):
mostly just reviewed for my own education
